TranscriptAnnotator.pm

```diff
@@ -0,0 +1,261 @@
+             the filename without extensions). Set to 0 to avoid any prefix.
+   trim:     Trim whitespaces from both ends of each column (default: 1).
+
+ The tabix and bgzip utilities must be installed in your path to read the
+ tabix-indexed annotation file: check https://github.com/samtools/htslib.git for
+ installation instructions.
+
+=cut
+
+package TranscriptAnnotator;
+
+use strict;
+use warnings;
+
+use File::Basename;
+
+use Bio::EnsEMBL::Variation::Utils::Sequence qw(get_matched_variant_alleles);
+
+use Bio::EnsEMBL::Variation::Utils::BaseVepTabixPlugin;
+use base qw(Bio::EnsEMBL::Variation::Utils::BaseVepTabixPlugin);
+
+sub _basename_without_ext {
+  my $file = basename shift;
+  $file =~ s/\.[^.]*?(.gz)?$//g;
+  return $file;
+}
+
+sub _trim_whitespaces {
+  my $text = shift;
+  $text =~ s/^\s+|\s+$//g;
+  return $text;
+};
+
+sub _get_colnames {
+  my $self = shift;
+
+  # Open file header
+  open IN, "tabix -f -h " . $self->{_files}[0] . " 1:1-1 |"
+    or die "ERROR: cannot open tabix file for " . $self->{_files}[0];
+
+  # Get last line
+  my $last;
+  $last = $_ while <IN>;
+  $last =~ s/(^#|\n$)//g;
+  close IN;
+
+  # Parse column names from header
+  my @cols = split /\t/, $last;
+  @cols = map { _trim_whitespaces $_ } @cols if $self->{trim};
+  @cols = splice @cols, 5;
+  return @cols;
+}
+
+sub _get_selected_cols {
+  my $self = shift;
+  my @colnames = @_;
+
+  my @cols;
+  my $param_hash = $self->params_to_hash();
+  if ( $param_hash->{cols} ) {
+    # Check if user-selected columns are valid
+    my @invalid_cols;
+    for ( split /:/, $param_hash->{cols} ) {
+      if ($_ ~~ @colnames) {
+        push @cols, $_;
+      } else {
+        push @invalid_cols, $_;
+      }
+    }
+
+    # Warn about invalid user-selected columns
+    my $msg_valid_cols = "Valid column names: " . join(", ", @colnames) . "\n";
+    if (not @cols) {
+      die "ERROR: no valid columns selected. " . $msg_valid_cols;
+    } elsif (@invalid_cols) {
+      my $file = $self->{_files}[0];
+      warn "WARNING: columns " . join(", ", @invalid_cols) .
+           " not found in $file. " . $msg_valid_cols;
+    }
+  } else {
+    @cols = @colnames;
+  }
+  return @cols;
+}
+
+sub new {
+  my $class = shift;
+  my $self = $class->SUPER::new(@_);
+  $self->get_user_params();
+
+  my $param_hash = $self->params_to_hash();
+  $self->{trim} = defined($param_hash->{trim}) ? $param_hash->{trim} : 1;
+
+  # Check file
+  my $file = $param_hash->{file};
+  die "\nERROR: No file specified\nTry using 'file=path/to/file.txt.gz'\n"
+     unless defined($file);
+  $self->add_file($file);
+
+  # Prepare column names
+  my @colnames = $self->_get_colnames();
+  my @cols     = $self->_get_selected_cols(@colnames);
+
+  # Add prefix to column names
+  unless (defined $param_hash->{prefix} && $param_hash->{prefix} eq 0) {
+    $self->{prefix} = $param_hash->{prefix} || _basename_without_ext($file) . "_";
```

Comment:
It should be:
`$self->{prefix} = ($param_hash->{prefix} || _basename_without_ext($file)) . "_";`
Otherwise '_' it's only added before the filename, not the prefix.

Reply:
Yes, fixed!